### ISSUE TITLE
Provide indeterminate state for checkboxes

### ISSUE THREAD

**Issue by SamSchott** [beeware/toga#1241]
**Is your feature request related to a problem? Please describe.**
The `Switch` widget currently support only two states: on and off. This makes sense on mobile UIs where it is actually rendered as a switch. On desktops, where it is rendered as a checkbox, providing a third indeterminate state is useful for certain applications and is commonly supported by the underlying native widgets.

The main use case is a hierarchy of selection or options where some children are checked and some are not. The parent would in this case be in an indeterminate state. For example:

![image](https://user-images.githubusercontent.com/20974290/113506788-c5902000-953e-11eb-919b-60c47434102d.png)

**Describe the solution you'd like**
It would be good to provide this option where supported by the underlying widget.

**Describe alternatives you've considered**
Alternatively, it could make sense to provide separate `Switch` and `Checkbox` widgets. This is supported through equivalent native widgets at on most platforms:

|      | macOS | Gnome | WinForms | iOS | Android |
| -- | --------| ------- | ----------| --- | ---------|
| switch | NSSwitch | GtkSwitch | -- | UISwitch | Switch |
| checkbox | NSButton | GtkCheckButton | CheckBox | -- | CheckBox |

**Comment by freakboy3742** [comment]
I've definitely seen the type of widget you're describing - although, as you note, it's usually in the context of a Tree layout, or at least where there's a hierarchy, and you're trying to reflect child states.

As for how to represent this in the Toga API - that I'm not so sure about.

The goal of Toga's API isn't to just wrap the underlying widgets. We're also trying to capture the high level usage of those widgets. Although we do have a "switch" widget, it shouldn't be thought of as "a wrapper around a switch", but "the platform dependent way to represent a boolean". On iOS, the HIG says booleans should be switches, because the touch-s; on macOS and Windows, they're checkboxes. The actual appearance of the widget is a platform detail; the use case is "show the user a thing that can be turned on and off".

So - if we were to accomodate an "unknown" state, rather than a "Switch" and "Checkbox" as two different physical manifestations of widgets, we would be adding either:

1) An "allow unknown state" option to Switch, or 
2) A new "tristate" widget type (although I'm not sure what you'd call this widget...)

My gut reaction is that of those two, (1) is more preferable. On platforms that only have checkbox, nothing much will change (other than a third state being possible); platforms that have both "switch" and "checkbox" could use the "allow unknown" option to determine which physical widget is used. The big question is what we do on iOS, where there isn't a native concept of a checkbox, but a switch doesn't allow for an "unknown" state. 

A completely different approach would be to look at the larger scale problem. This "tristate" widget only comes up when there's a hierarchy, and the parent needs to reflect the mixed state of its children. This might indicate that the right course of action is to only expose this sort of feature as an internal feature of a tree-based control (either the literal Tree widget as we have it right now, or a "selection hierarchy tree" - i.e., like a radio button group, but where the options are all checkboxes, and the state of the parent control the state of the children). In that latter context, the native iOS rendering won't be a checkbox *or* a switch - it will be a hierarchical collection of lists, with a navigation element that allows you to "dig into" the child lists. This would allow us to capture the actual use case for the tristate button, without actually having a literal tristate button.

Thoughts?

**Comment by SamSchott** [comment]
I agree that having a single widget to represent a boolean state would be cleaner and that the use cases for tristate checkboxes are very niche. The idea of providing such tristate widgets only where they would likely be required and not exposing them as a separate widget would certainly bypass issues with API design. Widgets like `toga.Tree` are only available on desktops anyways. 

However, there might also be other arguments for exposing a separate checkbox widget and implementing the switch as a "switch" everywhere. Your comment about the HIG had me look into the platforms a bit:

* **iOS**: has a very definitive preference for dual-state switches. Checkbox-like functionality is occasionally exposed in the settings app through check marks (for example Settings > FaceTime > You Can Be Reached By fateTime At)
*  **Android**: [specifically asks](https://material.io/components/checkboxes#usage) developers to use checkboxes instead of switches in some circumstances.
* **macOS**: [allows switches](https://developer.apple.com/design/human-interface-guidelines/macos/buttons/switches/) for major settings outside of frames, as used in Settings > Notifications.
* **Windows**: [specifically requests](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/toggles) the use of switches instead of checkboxes in some contexts.
* **Gnome**: has similar preferences for [switches](https://developer.gnome.org/hig/stable/switches.html.en) and [checkboxes](https://developer.gnome.org/hig/stable/check-boxes.html.en) depending on the context.

It appears that most platform HIG prefer switches over checkboxes in specific contexts. So if we really want to be HIG compliant, I actually don't see a away around providing both...

**Comment by freakboy3742** [comment]
Closing as a duplicate of #2225.
